### ISSUE TITLE
[BUG] Cannot allocate memory!: Not enough space

### ISSUE THREAD

**Issue by RainbowMiner** [sp-hash/TeamBlackMiner#9]
The miner v1.0 crashes regularly on a 6xGTX1070 / Windows 10 / CUDA 11.3 (Nvidia v466.74) mining rig while mining Ethash to Nicehash. DRAM is 8GB, pagefile/virtual memory is set to 120GB (120000). Sometimes it crashes without the cuda runtime error, but always with twice `Critical error. Cannot allocate memory!: Not enough space`

```
[1;36m     3m[0m [1m[2021-09-10 15:22:50.[0m[0m224[0m[1m][0m [31mC:\gc\TBMiner\source\sp_cuda\cudaminer.cpp:cuda_search():349: A cuda runtime error occured: an illegal memory access was encountered[0m
[1;36m     3m[0m [1m[2021-09-10 15:22:50.[0m[0m224[0m[1m][0m [36mShutting down threads[0m
[1;36m     3m[0m [1m[2021-09-10 15:22:50.[0m[0m224[0m[1m][0m [31mC:\gc\TBMiner\source\sp_cuda\cudaminer.cpp:cuda_search():349: A cuda runtime error occured: an illegal memory access was encountered[0m
[1;36m     3m[0m [1m[2021-09-10 15:22:50.[0m[0m224[0m[1m][0m [36mShutting down threads[0m
[1;36m     3m[0m [1m[2021-09-10 15:22:50.[0m[0m240[0m[1m][0m [31mC:\gc\TBMiner\source\sp_cuda\cudaminer.cpp:cuda_search():349: A cuda runtime error occured: an illegal memory access was encountered[0m
Critical error. Cannot allocate memory!: Not enough space
Critical error. Cannot allocate memory!: Not enough space
```

**Comment by RainbowMiner** [comment]
Same happens on 6xGTX1070 / Windows 10 / CUDA 11.4 (Nvidia v471.41), mining Ethash to Ethermine.
```
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m308[0m[1m][0m [31mC:\gc\TBMiner\source\sp_cuda\cudaminer.cpp:cuda_search():349: A cuda runtime error occured: unspecified launch failure[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m308[0m[1m][0m [36mShutting down threads[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m323[0m[1m][0m [33mStratum thread exited[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m323[0m[1m][0m [33mControl thread exited[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m339[0m[1m][0m [31mC:\gc\TBMiner\source\sp_cuda\cudaminer.cpp:cuda_search():349: A cuda runtime error occured: unspecified launch failure[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m339[0m[1m][0m [36mShutting down threads[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m339[0m[1m][0m [31mC:\gc\TBMiner\source\sp_cuda\cudaminer.cpp:cuda_search():349: A cuda runtime error occured: unspecified launch failure[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m339[0m[1m][0m [36mShutting down threads[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m339[0m[1m][0m [31mC:\gc\TBMiner\source\sp_cuda\cudaminer.cpp:cuda_search():349: A cuda runtime error occured: unspecified launch failure[0m
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m339[0m[1m][0m [36mShutting down threads[0m
Critical error. Cannot allocate memory!: Not enough space
Critical error. Cannot allocate memory!: Not enough space
[1;36m     3m[0m [1m[2021-09-10 16:22:14.[0m[0m339[0m[1m][0m [31mC:\gc\TBMiner\source\sp_cuda\cudaminer.cpp:cu
```

**Comment by sp-hash** [comment]
Add more virtual memory 32gb

**Comment by RainbowMiner** [comment]
I have set to 160gb, now. But it still happens. This time etchash on Ethermine, after 2-3 minutes runtime
```
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [1mID	BOARD	TEMP	FAN	CORE	MEM	WATT	kW/h	COST/h[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [36mGPU0	1070	63	43	1956	4404	135	0.14	inf[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [36mGPU1	1070	67	62	1939	4404	117	0.12	inf[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [36mGPU2	1070	68	66	1911	4404	135	0.14	inf[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [36mGPU3	1070	68	66	1863	4404	134	0.13	inf[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [36mGPU4	1070	63	50	1936	4404	132	0.13	inf[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [36mGPU5	1070	65	54	1927	4404	135	0.14	inf[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [1;36m						788	0.79	inf[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [0m[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [1mID	BOARD	TYPE	HASHRATE/W	HASHRATE	AVERAGE		SHARES[0m
[1;36m     2m[0m [1m[2021-09-10 21:08:35.[0m[0m811[0m[1m][0m [36mGPU0	1070	Cuda	231.78 kH/s	31.29 MH/s	31.29 MH/s	0/0 (100.00)[0m
Critical error. Cannot allocate memory!: Not enough space
Critical error. Cannot allocate memory!: Not enough space
```

**Comment by sp-hash** [comment]
Did you try v1.01? Which driver and exe are you using?

**Comment by RainbowMiner** [comment]
Yes, v1.01 - Nvidia v471.41 thus CUDA11.4
It seems to happen mostly during Etchash mining.

**Comment by sp-hash** [comment]
Use gpu-z and check if you have enough memory. Perhaps a process is not shut down correctly or you didn't wait long enough when killing the process. What's the launch string of the miner`?

Btw, or no colours use the option:

-I --no-ansi

**Comment by RainbowMiner** [comment]
I have checked with gpu-z - there is more than enough memory all the time. So it's something different.
Could that crash be related to overclocking or damaged DAGs? This morning, the miner (on ethash) ran long time without crash, but it seem to have lost one GPU after 5 minutes `GPU 3: has timed out`. From that time on, it reported 0 H/s for that specific GPU.
Also, it seems, that the etchash does crash the miner more often.

Thanks for the `--no-ansi` hint :) makes the logs much more readable.

**Comment by sp-hash** [comment]
Thanks for the report. The error is because the miner doesn't shut down properly. 
If the gpu crash, the hashrate will go to 0. This is probobly because the clock is to high, or the rig is using too much power. 

To make sure the miner exits properly run it in a loop with a timeout.
:loop
	TBMiner --algo ethash --hostname s1.solopool.org --port 8002 --wallet %ETH_WALLET% --worker_name %computername%
	timeout 10
goto :loop

**Comment by sp-hash** [comment]
v1.02 will have a small code change that might help here.

**Comment by sp-hash** [comment]
v1.02 beta has been released with a possible fix

**Comment by RainbowMiner** [comment]
Thank you! I have updated the RainbowMiner library with the v1.02. Now I have to wait some hours, until one of my rigs isn't rented any more. I'll report back as soon as the new version gets benchmarked.

**Comment by sp-hash** [comment]
fixed in v1.02
